// @module-federation/managers/dist/index.mjs
import { BasicPluginOptionsManager } from "./BasicPluginOptionsManager.mjs";
import { ContainerManager } from "./ContainerManager.mjs";
import { PKGJsonManager } from "./PKGJsonManager.mjs";
import { RemoteManager } from "./RemoteManager.mjs";
import { SharedManager } from "./SharedManager.mjs";
import { UNKNOWN_MODULE_NAME } from "./constant.mjs";
import * as __WEBPACK_EXTERNAL_MODULE__utils_mjs_25ece7d1__ from "./utils.mjs";
import * as __WEBPACK_EXTERNAL_MODULE__types_mjs_4ad50757__ from "./types.mjs";

;// CONCATENATED MODULE: external "./BasicPluginOptionsManager.mjs"

;// CONCATENATED MODULE: external "./ContainerManager.mjs"

;// CONCATENATED MODULE: external "./PKGJsonManager.mjs"

;// CONCATENATED MODULE: external "./RemoteManager.mjs"

;// CONCATENATED MODULE: external "./SharedManager.mjs"

;// CONCATENATED MODULE: external "./constant.mjs"

;// CONCATENATED MODULE: external "./utils.mjs"

;// CONCATENATED MODULE: external "./types.mjs"

;// CONCATENATED MODULE: ./src/index.ts









export { BasicPluginOptionsManager, ContainerManager, PKGJsonManager, RemoteManager, SharedManager, UNKNOWN_MODULE_NAME, __WEBPACK_EXTERNAL_MODULE__types_mjs_4ad50757__ as types, __WEBPACK_EXTERNAL_MODULE__utils_mjs_25ece7d1__ as utils };
