// @module-federation/webpack-bundler-runtime/dist/index.js
import { attachShareScopeMap } from "./attachShareScopeMap.js";
import { remotes } from "./remotes.js";
import { consumes } from "./consumes.js";
import { initializeSharing } from "./initializeSharing.js";
import { installInitialConsumes } from "./installInitialConsumes.js";
import { initContainerEntry } from "./initContainerEntry.js";
import { init } from "./init.js";
import { getSharedFallbackGetter } from "./getSharedFallbackGetter.js";
import * as runtime from "@module-federation/runtime";

//#region src/index.ts
const federation = {
	runtime,
	instance: void 0,
	initOptions: void 0,
	bundlerRuntime: {
		remotes,
		consumes,
		I: initializeSharing,
		S: {},
		installInitialConsumes,
		initContainerEntry,
		init,
		getSharedFallbackGetter
	},
	attachShareScopeMap,
	bundlerRuntimeOptions: {}
};
const instance = federation.instance;
const initOptions = federation.initOptions;
const bundlerRuntime = federation.bundlerRuntime;
const bundlerRuntimeOptions = federation.bundlerRuntimeOptions;

//#endregion
export { attachShareScopeMap, bundlerRuntime, bundlerRuntimeOptions, federation as default, initOptions, instance, runtime };
