// @module-federation/enhanced/dist/src/lib/container/ContainerExposedDependency.js
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });
const require_runtime = require('../../_virtual/_rolldown/runtime.js');
let _module_federation_sdk_normalize_webpack_path = require("@module-federation/sdk/normalize-webpack-path");

//#region src/lib/container/ContainerExposedDependency.ts
const makeSerializable = require((0, _module_federation_sdk_normalize_webpack_path.normalizeWebpackPath)("webpack/lib/util/makeSerializable"));
const { dependencies } = require((0, _module_federation_sdk_normalize_webpack_path.normalizeWebpackPath)("webpack"));
var ContainerExposedDependency = class extends dependencies.ModuleDependency {
	/**
	* @param {string} exposedName public name
	* @param {string} request request to module
	*/
	constructor(exposedName, request) {
		super(request);
		this.exposedName = exposedName;
		this.request = request;
	}
	get type() {
		return "container exposed";
	}
	get category() {
		return "esm";
	}
	/**
	* @returns {string | null} an identifier to merge equal requests
	*/
	getResourceIdentifier() {
		return `exposed dependency ${this.exposedName}=${this.request}`;
	}
	/**
	* @param {ObjectSerializerContext} context context
	*/
	serialize(context) {
		context.write(this.exposedName);
		super.serialize(context);
	}
	/**
	* @param {ObjectDeserializerContext} context context
	*/
	deserialize(context) {
		this.exposedName = context.read();
		super.deserialize(context);
	}
};
makeSerializable(ContainerExposedDependency, "enhanced/lib/container/ContainerExposedDependency");

//#endregion
exports.default = ContainerExposedDependency;
