// @module-federation/enhanced/dist/src/lib/container/FallbackDependency.js
'use strict';

Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });
const require_runtime = require('../../_virtual/_rolldown/runtime.js');
let _module_federation_sdk_normalize_webpack_path = require("@module-federation/sdk/normalize-webpack-path");

//#region src/lib/container/FallbackDependency.ts
const makeSerializable = require((0, _module_federation_sdk_normalize_webpack_path.normalizeWebpackPath)("webpack/lib/util/makeSerializable"));
const { Dependency } = require((0, _module_federation_sdk_normalize_webpack_path.normalizeWebpackPath)("webpack"));
var FallbackDependency = class FallbackDependency extends Dependency {
	/**
	* @param {string[]} requests requests
	*/
	constructor(requests) {
		super();
		this.requests = requests;
	}
	/**
	* @returns {string | null} an identifier to merge equal requests
	*/
	getResourceIdentifier() {
		return `fallback ${this.requests.join(" ")}`;
	}
	get type() {
		return "fallback";
	}
	get category() {
		return "esm";
	}
	/**
	* @param {ObjectSerializerContext} context context
	*/
	serialize(context) {
		const { write } = context;
		write(this.requests);
		super.serialize(context);
	}
	static deserialize(context) {
		const { read } = context;
		const obj = new FallbackDependency(read());
		obj.deserialize(context);
		return obj;
	}
};
makeSerializable(FallbackDependency, "enhanced/lib/container/FallbackDependency");

//#endregion
exports.default = FallbackDependency;
