// @module-federation/dts-plugin/dist/consumeTypes-DxGMNavh.js
const require_Action = require('./Action-CzhPMw2i.js');
const require_expose_rpc = require('./expose-rpc-CQq6zYK-.js');
let path = require("path");
path = require_Action.__toESM(path);
let crypto = require("crypto");
let child_process = require("child_process");
child_process = require_Action.__toESM(child_process);
let url = require("url");
let process$1 = require("process");
process$1 = require_Action.__toESM(process$1);

//#region src/core/rpc/rpc-error.ts
var RpcExitError = class extends Error {
	constructor(message, code, signal) {
		super(message);
		this.code = code;
		this.signal = signal;
		this.name = "RpcExitError";
	}
};

//#endregion
//#region src/core/rpc/wrap-rpc.ts
function createControlledPromise() {
	let resolve = () => void 0;
	let reject = () => void 0;
	return {
		promise: new Promise((aResolve, aReject) => {
			resolve = aResolve;
			reject = aReject;
		}),
		resolve,
		reject
	};
}
function wrapRpc(childProcess, options) {
	return (async (...args) => {
		if (!childProcess.send) throw new Error(`Process ${childProcess.pid} doesn't have IPC channels`);
		else if (!childProcess.connected) throw new Error(`Process ${childProcess.pid} doesn't have open IPC channels`);
		const { id, once } = options;
		const { promise: resultPromise, resolve: resolveResult, reject: rejectResult } = createControlledPromise();
		const { promise: sendPromise, resolve: resolveSend, reject: rejectSend } = createControlledPromise();
		const handleMessage = (message) => {
			if (message?.id === id) {
				if (message.type === require_expose_rpc.RpcGMCallTypes.RESOLVE) resolveResult(message.value);
				else if (message.type === require_expose_rpc.RpcGMCallTypes.REJECT) rejectResult(message.error);
			}
			if (once && childProcess?.kill) childProcess.kill("SIGTERM");
		};
		const handleClose = (code, signal) => {
			rejectResult(new RpcExitError(code ? `Process ${childProcess.pid} exited with code ${code}${signal ? ` [${signal}]` : ""}` : `Process ${childProcess.pid} exited${signal ? ` [${signal}]` : ""}`, code, signal));
			removeHandlers();
		};
		const removeHandlers = () => {
			childProcess.off("message", handleMessage);
			childProcess.off("close", handleClose);
		};
		if (once) childProcess.once("message", handleMessage);
		else childProcess.on("message", handleMessage);
		childProcess.on("close", handleClose);
		childProcess.send({
			type: require_expose_rpc.RpcGMCallTypes.CALL,
			id,
			args
		}, (error) => {
			if (error) {
				rejectSend(error);
				removeHandlers();
			} else resolveSend(void 0);
		});
		return sendPromise.then(() => resultPromise);
	});
}

//#endregion
//#region src/core/rpc/rpc-worker.ts
const FEDERATION_WORKER_DATA_ENV_KEY = "VMOK_WORKER_DATA_ENV";
function createRpcWorker(modulePath, data, memoryLimit, once) {
	const options = {
		env: {
			...process$1.env,
			[FEDERATION_WORKER_DATA_ENV_KEY]: JSON.stringify(data || {})
		},
		stdio: [
			"inherit",
			"inherit",
			"inherit",
			"ipc"
		],
		serialization: "advanced"
	};
	if (memoryLimit) options.execArgv = [`--max-old-space-size=${memoryLimit}`];
	let childProcess, remoteMethod;
	const id = (0, crypto.randomUUID)();
	return {
		connect(...args) {
			if (childProcess && !childProcess.connected) {
				childProcess.send({
					type: require_expose_rpc.RpcGMCallTypes.EXIT,
					id
				});
				childProcess = void 0;
				remoteMethod = void 0;
			}
			if (!childProcess?.connected) {
				childProcess = child_process.fork(modulePath, options);
				remoteMethod = wrapRpc(childProcess, {
					id,
					once
				});
			}
			if (!remoteMethod) return Promise.reject(/* @__PURE__ */ new Error("Worker is not connected - cannot perform RPC."));
			return remoteMethod(...args);
		},
		terminate() {
			try {
				if (childProcess.connected) childProcess.send({
					type: require_expose_rpc.RpcGMCallTypes.EXIT,
					id
				}, (err) => {
					if (err) console.error("Error sending message:", err);
				});
			} catch (error) {
				if (error.code === "EPIPE") console.error("Pipe closed before message could be sent:", error);
				else console.error("Unexpected error:", error);
			}
			childProcess = void 0;
			remoteMethod = void 0;
		},
		get connected() {
			return Boolean(childProcess?.connected);
		},
		get process() {
			return childProcess;
		},
		get id() {
			return id;
		}
	};
}
function getRpcWorkerData() {
	return JSON.parse(process$1.env[FEDERATION_WORKER_DATA_ENV_KEY] || "{}");
}

//#endregion
//#region src/core/rpc/index.ts
var rpc_exports = /* @__PURE__ */ require_Action.__exportAll({
	RpcExitError: () => RpcExitError,
	RpcGMCallTypes: () => require_expose_rpc.RpcGMCallTypes,
	createRpcWorker: () => createRpcWorker,
	exposeRpc: () => require_expose_rpc.exposeRpc,
	getRpcWorkerData: () => getRpcWorkerData,
	wrapRpc: () => wrapRpc
});

//#endregion
//#region src/core/lib/DtsWorker.ts
const __filename$1 = (0, url.fileURLToPath)(require("url").pathToFileURL(__filename).href);
const __dirname$1 = path.default.dirname(__filename$1);
const __extname = path.default.extname(__filename$1);
var DtsWorker = class {
	constructor(options) {
		this._options = require_expose_rpc.cloneDeepOptions(options);
		this.removeUnSerializationOptions();
		this.rpcWorker = createRpcWorker(path.default.resolve(__dirname$1, `./fork-generate-dts${__extname}`), {}, void 0, true);
		this._res = this.rpcWorker.connect(this._options);
	}
	removeUnSerializationOptions() {
		if (this._options.remote?.moduleFederationConfig?.manifest) delete this._options.remote?.moduleFederationConfig?.manifest;
		if (this._options.host?.moduleFederationConfig?.manifest) delete this._options.host?.moduleFederationConfig?.manifest;
	}
	get controlledPromise() {
		const ensureChildProcessExit = () => {
			try {
				const pid = this.rpcWorker.process?.pid;
				const rootPid = process.pid;
				if (pid && rootPid !== pid) process.kill(pid, 0);
			} catch (error) {
				if (require_expose_rpc.isDebugMode()) console.error(error);
			}
		};
		return Promise.resolve(this._res).then(() => {
			this.exit();
			ensureChildProcessExit();
		}).catch((err) => {
			if (require_expose_rpc.isDebugMode()) console.error(err);
			ensureChildProcessExit();
		});
	}
	exit() {
		try {
			this.rpcWorker?.terminate();
		} catch (err) {
			if (require_expose_rpc.isDebugMode()) console.error(err);
		}
	}
};

//#endregion
//#region src/core/lib/generateTypesInChildProcess.ts
async function generateTypesInChildProcess(options) {
	return new DtsWorker(options).controlledPromise;
}

//#endregion
//#region src/core/lib/consumeTypes.ts
async function consumeTypes(options) {
	await new (require_expose_rpc.getDTSManagerConstructor(options.host?.implementation))(options).consumeTypes();
}

//#endregion
Object.defineProperty(exports, 'DtsWorker', {
  enumerable: true,
  get: function () {
    return DtsWorker;
  }
});
Object.defineProperty(exports, 'consumeTypes', {
  enumerable: true,
  get: function () {
    return consumeTypes;
  }
});
Object.defineProperty(exports, 'createRpcWorker', {
  enumerable: true,
  get: function () {
    return createRpcWorker;
  }
});
Object.defineProperty(exports, 'generateTypesInChildProcess', {
  enumerable: true,
  get: function () {
    return generateTypesInChildProcess;
  }
});
Object.defineProperty(exports, 'rpc_exports', {
  enumerable: true,
  get: function () {
    return rpc_exports;
  }
});
