// @module-federation/dts-plugin/dist/core.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
require('./Broker-DRFgFvXI.js');
const require_expose_rpc = require('./expose-rpc-CQq6zYK-.js');
const require_consumeTypes = require('./consumeTypes-DxGMNavh.js');

exports.DTSManager = require_expose_rpc.DTSManager;
exports.DtsWorker = require_consumeTypes.DtsWorker;
exports.HOST_API_TYPES_FILE_NAME = require_expose_rpc.HOST_API_TYPES_FILE_NAME;
exports.REMOTE_ALIAS_IDENTIFIER = require_expose_rpc.REMOTE_ALIAS_IDENTIFIER;
exports.REMOTE_API_TYPES_FILE_NAME = require_expose_rpc.REMOTE_API_TYPES_FILE_NAME;
exports.consumeTypes = require_consumeTypes.consumeTypes;
exports.generateTypes = require_expose_rpc.generateTypes;
exports.generateTypesInChildProcess = require_consumeTypes.generateTypesInChildProcess;
exports.getDTSManagerConstructor = require_expose_rpc.getDTSManagerConstructor;
exports.isTSProject = require_expose_rpc.isTSProject;
exports.retrieveHostConfig = require_expose_rpc.retrieveHostConfig;
exports.retrieveMfTypesPath = require_expose_rpc.retrieveMfTypesPath;
exports.retrieveOriginalOutDir = require_expose_rpc.retrieveOriginalOutDir;
exports.retrieveRemoteConfig = require_expose_rpc.retrieveRemoteConfig;
exports.retrieveTypesAssetsInfo = require_expose_rpc.retrieveTypesAssetsInfo;
exports.retrieveTypesZipPath = require_expose_rpc.retrieveTypesZipPath;
Object.defineProperty(exports, 'rpc', {
  enumerable: true,
  get: function () {
    return require_consumeTypes.rpc_exports;
  }
});
exports.validateOptions = require_expose_rpc.validateOptions;
