// @module-federation/enhanced/dist/src/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('./_virtual/_rolldown/runtime.js');
const require_wrapper_ModuleFederationPlugin = require('./wrapper/ModuleFederationPlugin.js');
const require_wrapper_ContainerReferencePlugin = require('./wrapper/ContainerReferencePlugin.js');
const require_wrapper_SharePlugin = require('./wrapper/SharePlugin.js');
const require_wrapper_ContainerPlugin = require('./wrapper/ContainerPlugin.js');
const require_wrapper_ConsumeSharedPlugin = require('./wrapper/ConsumeSharedPlugin.js');
const require_wrapper_ProvideSharedPlugin = require('./wrapper/ProvideSharedPlugin.js');
const require_wrapper_FederationModulesPlugin = require('./wrapper/FederationModulesPlugin.js');
const require_wrapper_FederationRuntimePlugin = require('./wrapper/FederationRuntimePlugin.js');
const require_wrapper_AsyncBoundaryPlugin = require('./wrapper/AsyncBoundaryPlugin.js');
const require_wrapper_HoistContainerReferencesPlugin = require('./wrapper/HoistContainerReferencesPlugin.js');
const require_wrapper_TreeShakingSharedPlugin = require('./wrapper/TreeShakingSharedPlugin.js');
const require_lib_container_options = require('./lib/container/options.js');
let _module_federation_sdk = require("@module-federation/sdk");

//#region src/index.ts
const lazyRequire = (id) => module.require(id);
const dependencies = { get ContainerEntryDependency() {
	return lazyRequire("./lib/container/ContainerEntryDependency").default;
} };
const container = { get ContainerEntryModule() {
	return lazyRequire("./lib/container/ContainerEntryModule").default;
} };

//#endregion
exports.AsyncBoundaryPlugin = require_wrapper_AsyncBoundaryPlugin.default;
exports.ConsumeSharedPlugin = require_wrapper_ConsumeSharedPlugin.default;
exports.ContainerPlugin = require_wrapper_ContainerPlugin.default;
exports.ContainerReferencePlugin = require_wrapper_ContainerReferencePlugin.default;
exports.FederationModulesPlugin = require_wrapper_FederationModulesPlugin.default;
exports.FederationRuntimePlugin = require_wrapper_FederationRuntimePlugin.default;
exports.HoistContainerReferencesPlugin = require_wrapper_HoistContainerReferencesPlugin.default;
exports.ModuleFederationPlugin = require_wrapper_ModuleFederationPlugin.default;
exports.PLUGIN_NAME = require_wrapper_ModuleFederationPlugin.PLUGIN_NAME;
exports.ProvideSharedPlugin = require_wrapper_ProvideSharedPlugin.default;
exports.SharePlugin = require_wrapper_SharePlugin.default;
exports.TreeShakingSharedPlugin = require_wrapper_TreeShakingSharedPlugin.default;
exports.container = container;
Object.defineProperty(exports, 'createModuleFederationConfig', {
  enumerable: true,
  get: function () {
    return _module_federation_sdk.createModuleFederationConfig;
  }
});
exports.dependencies = dependencies;
exports.parseOptions = require_lib_container_options.parseOptions;
